include/deal.II/base/aligned_vector.h

```diff
@@ -363,12 +395,12 @@ namespace internal
           {
             // initialize memory (copy construct by placement new), and
             // destruct the source
-            new (&destination_[i]) T(source_[i]);
+            new (&destination_[i]) T(std::move(source_[i]));
             source_[i].~T();
           }
       else
         for (std::size_t i=begin; i<end; ++i)
-          new (&destination_[i]) T(source_[i]);
+          new (&destination_[i]) T(std::move(source_[i]));
```

Comment:
Do you need this change for compiling code? Despite the name `AlignedVectorMove` name of the class, this is the copy path. Doesn't `std::move` delete the old content and thus leave some memory in an undefined state? Can you run the test `tests/base/aligned_vector_02.debug` through valgrind?

Reply:
Good catch indeed -- the test failed because of this.

The standard doesn't really say what a move constructor or move operator does to the source object. But it must leave it in a state so that one can reasonably call a destructor on it, so all examples I've seen make sure that the source object is somehow left in a state where it has lost its resources but that that corresponds to an "empty" kind of state. So you should never get anything that valgrind catches.

To give an example, if you move construct a std::vector, you could think that now both source and destination point to the same memory location. That would be a standards conforming implementation of the operation, but it will of course lead to memory corruption once the first of the two objects is destroyed -- and so that's not how any reasonable person would implement things. Rather, reasonable programmers would set the pointer and size in the source std::vector to zero :-)